source/plugins/ruby/extension.rb

```diff
@@ -91,17 +109,59 @@ def get_stream_mapping()
      return dataTypeToStreamIdMap
   end
 
+  def get_namedpipe_mapping()
+    dataTypeToNamedPipeMap = Hash.new
+    begin
+     taggedAgentData = get_extension_configs(true)
+     if !taggedAgentData.nil? && !taggedAgentData.empty?
+      extensionConfigurations = taggedAgentData["extensionConfigurations"]
+      outputStreamDefinitions = taggedAgentData["outputStreamDefinitions"]
+      if !extensionConfigurations.nil? && !extensionConfigurations.empty?
+        extensionConfigurations.each do |extensionConfig|
+          outputStreams = extensionConfig["outputStreams"]
+          if !outputStreams.nil? && !outputStreams.empty?
+            outputStreams.each do |datatypeId, streamId|
+            dataTypeToNamedPipeMap[datatypeId] = outputStreamDefinitions[streamId]["namedPipe"]
+            end
+          else
+            $log.warn("Extension::get_namedpipe_mapping::received outputStreams or outputStreamDefinitions is either nil or empty")
+          end
+        end
+      else
+        $log.warn("Extension::get_namedpipe_mapping::received extensionConfigurations either nil or empty")
+      end
+     else
+      $log.warn("Extension::get_namedpipe_mapping::received taggedAgentData either nil or empty")
+     end
+    rescue => errorStr
+     $log.warn("Extension::get_namedpipe_mapping failed: #{errorStr}")
+     ApplicationInsightsUtility.sendExceptionTelemetry(errorStr)
+    end
+    return dataTypeToNamedPipeMap
+ end
+
+
   private
-  def get_extension_configs()
+  def get_extension_configs(getTaggedAgentData = false)
     extensionConfigurations = []
     begin
-      clientSocket = UNIXSocket.open(Constants::ONEAGENT_FLUENT_SOCKET_NAME)
       requestId = SecureRandom.uuid.to_s
       requestBodyJSON = { "Request" => "AgentTaggedData", "RequestId" => requestId, "Tag" => Constants::CI_EXTENSION_NAME, "Version" => Constants::CI_EXTENSION_VERSION }.to_json
-      requestBodyMsgPack = requestBodyJSON.to_msgpack
-      clientSocket.write(requestBodyMsgPack)
-      clientSocket.flush
-      resp = clientSocket.recv(Constants::CI_EXTENSION_CONFIG_MAX_BYTES)
+      if !@@isWindows.nil? && @@isWindows == false
+        clientSocket = UNIXSocket.open(Constants::ONEAGENT_FLUENT_SOCKET_NAME)
+        requestBodyMsgPack = requestBodyJSON.to_msgpack
+        clientSocket.write(requestBodyMsgPack)
+        clientSocket.flush
+        resp = clientSocket.recv(Constants::CI_EXTENSION_CONFIG_MAX_BYTES)
+      else
+        if !@clientNamedPipe
+          configPipe = "\\\\.\\pipe\\CAgentStream_CloudAgentInfo_AzureMonitorAgent"
+          @clientNamedPipe = File.open(configPipe, "w+")

```

Comment:
if the File.open fails for any reason, you will  have to set @clientNamedPipe is nil so that retry can happen.